### PR TITLE
fix(cdk/overlay): animations interrupted on repeat insertions

### LINKED ISSUE #24749: help(Overlay): Animation works only on first attachment ,
### What are you trying to do?

I'm trying to implement a generic auto completion.

My code is structured in following way

1. `dsAutocomplete` directive is top level directive 
2. `dsAutocompleteInputTrigger` is directive attached to `<input>` element
3. `dsAutoCompleteOrigin` is attached to element on whose drop down need to be anchored
4. `*dsAutocompletePanel` is a structural directive to define what is dropdown

Flow of actions is   

For opening 
`dsAutocompleteInputTrigger.focus` -> `dsAutocomplete.openPanel` -> `overlayRef.attach(TemplatePortal(dsAutocompletePanel))` as FlexibleConnected to `dsAutoCompleteOrigin`  

For closing 
`dsAutocompleteInputTrigger.focusout` -> `dsAutocomplete.closePanel` -> `overlayRef.detach()` 

when I add an animation directive any where in AutoComplete Panel it only works first time

### What troubleshooting steps have you tried?

Tried adding attach and detach in zone.run 

### Reproduction

follow instruction in following stackblitz
https://stackblitz.com/edit/angular-ivy-smmzcq?file=src%2Fapp%2Fapp.component.html

### Environment

- Angular:13.2
- CDK/Material:13.2
- Browser(s):Firefox , Chrome
- Operating System (e.g. Windows, macOS, Ubuntu):Windows 11

### PR DESCRIPTION
**Pull request by crisbeto**
When an overlay is detached, we remove it from the change detection tree and the DOM, but we don't destroy it completely so that it can be re-attached. Re-attachment is the same process, but in reverse: we re-add the element to the DOM and the change detection tree. The problem is that we were attaching the element to the change detection tree before re-inserting it into the DOM which caused the Angular animations package not to animate the element since it's not in the DOM yet.

These changes resolve the issue by attaching the element to the DOM first.

Fixes #24749.

**Note:** writing a unit test for this is tricky, because the content is never actually destroyed so we can't use lifecycle hooks for assertions.

### CODE REVIEW

**Review by jelbourn — APPROVED** [review]
LGTM
